relays/ethereum/src/ethereum_client.rs

```diff
@@ -150,6 +214,172 @@ async fn transaction_receipt(client: Client, hash: H256) -> (Client, Result<Rece
 	)
 }
 
+/// Returns best Substrate block that PoA chain knows of.
+pub async fn best_substrate_block(
+	client: Client,
+	contract_address: Address,
+) -> (Client, Result<SubstrateHeaderId, Error>) {
+	let (encoded_call, call_decoder) = bridge_contract::functions::best_known_header::call();
+	let call_request = bail_on_arg_error!(
+		to_value(CallRequest {
+			to: Some(contract_address),
+			data: Some(encoded_call.into()),
+		})
+		.map_err(|e| Error::RequestSerialization(e)),
+		client
+	);
+	let (client, call_result) =
+		bail_on_error!(call_rpc::<Bytes>(client, "eth_call", Params::Array(vec![call_request]),).await);
+	let (number, raw_hash) = match call_decoder.decode(&call_result.0) {
+		Ok((raw_number, raw_hash)) => (raw_number, raw_hash),
+		Err(error) => return (client, Err(Error::ResponseParseFailed(format!("{}", error)))),
```

Comment:
It's a bit out of scope now, but this pattern of returning `Client` is a bit annoying. Why do we need to own it?

Reply:
NVM, found why.